### PR TITLE
[Feat] Resolve manual_backward

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolves manual_optimization with DDP and new accelerator refactor.

@awaelchli @justusschock, when reviewed, it should be merged in the `accelerator-refactor-sharded` branch.

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -178,6 +181,15 @@ def set_world_ranks(self):
         self.world_size = self.num_nodes * self.num_processes
 
     def configure_ddp(self):
+
+        # todo: PyTorch 1.7.0 DDP introduces ``self.reducer._rebuild_buckets()``` breaking manual_optimization
+        if _PYTORCH_GREATER_EQUAL_THAN_1_7_0 and not self.lightning_module.automatic_optimization:
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp.py:186]: Do we have to make all these changes to ddp spawn as well?
Since this also uses ddp just with spawned processes...

**Review by justusschock — APPROVED** [review]
LGTM. Just one question.

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:186]: Oh yeah, you are right ! My bad.

**Comment by Borda** [comment]
@tchaton DONT MERGE is not a blocker, use WIP which is blocking merge

File: pytorch_lightning/overrides/base.py [hunk 2]
```diff
@@ -46,6 +46,13 @@ def forward(self, *inputs, **kwargs):
 
         if running_stage == RunningStage.TRAINING:
             output = self.module.training_step(*inputs, **kwargs)
+
+            # In manual_optimization, we need to prevent DDP reducer as
+            # it is done manually in ``LightningModule.manual_backward``
+            # `require_backward_grad_sync` will be reset
+            # ddp_plugin ``post_training_step`` hook
+            if not self.module.automatic_optimization:
+                self.module.trainer.model.require_backward_grad_sync = False
```
Comment (awaelchli) [inline pytorch_lightning/overrides/base.py:55]: I think this should not be here, looking at the call: `self.module.trainer.model` is the pytorch DDP wrapper.
So this can be added outside, maybe in the DDP training type plugin

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 3]
```diff
@@ -177,7 +180,19 @@ def set_world_ranks(self):
...
         self.world_size = self.num_nodes * self.num_processes
 
+    def pre_configure_ddp(self):
+        # todo: PyTorch 1.7.0 DDP introduces ``self.reducer._rebuild_buckets()``` breaking manual_optimization
+        if _PYTORCH_GREATER_EQUAL_THAN_1_7_0 and not self.lightning_module.automatic_optimization:
+            rank_zero_warn(
+                "From PyTorch 1.7.0, Lightning ``manual_optimization`` needs to set ``find_unused_parameters=True`` "
+                "to properly work with DDP."
+            )
+            self._ddp_kwargs["find_unused_parameters"] = True
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:190]: I don't get it, why do we need it in our refactor? I can't see this anywhere in 1.2 branch
Isn't the call to prepare_for_backward below enough?

File: pytorch_lightning/plugins/training_type/horovod.py [hunk 4]
```diff
@@ -116,6 +116,9 @@ def broadcast(self, obj: object, src: int = 0) -> object:
         obj = hvd.broadcast_object(obj, src)
         return obj
 
+    def post_backward(self, closure_loss: torch.Tensor, optimizer: Optimizer, opt_idx: int):
+        optimizer.synchronize()
+
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/horovod.py:121]: nice, then we should remove the hack in accelerator. There is this code:
```
if isinstance(self.training_type_plugin, HorovodPlugin):
            optimizer.synchronize()
```
we should remove it

File: pytorch_lightning/overrides/base.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/base.py:55]: nvm, although this looks suboptimal, @justusschock explained why it is necessary.

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp.py:190]: Justus explained it to me

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/horovod.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/horovod.py:121]: I removed it in this PR :)
